### PR TITLE
Cmake support - generator in configure.py

### PR DESCRIPTION
**Pull request by ksanderon**
Content:
* added cmake generator to configure.py
* added .astylerc
* updated .gitignore to not see generated cmakelists

~~TODO: check on platform(for instance on mac os x) which require "frameworks" if cmake list is generated correctly.~~ **edit: done**

I'm python noob, so please point me places where things could be made smarter/cleaner/better ;)

### CODE REVIEW

**Comment by webmaster128** [comment]
Just a comment unrelated to the quality of this PR: Caution before merging: this will clash with the changes in #962 even if git can merge it together. I can easily rebase #962 if this one goes in first. Otherwise this one needs to be updated.

File: configure.py [hunk 1]
```diff
@@ -1420,6 +1424,106 @@ def bakefile_test_sources(file, sources):
...
+    test_module = create_basic_rules(build_config.test_sources)
+    cli_module = create_basic_rules(build_config.cli_sources)
+
+    for module in using_mods:
+        update_rules(lib_module, module)
+        update_rules(test_module, module)
+        update_rules(cli_module, module)
+
+    f = open('CMakeLists.txt', 'w')
+    f.write('cmake_minimum_required(VERSION 3.0.0)\n')
```
Comment (webmaster128) [inline configure.py]: cmake 3.0 is a high requirement. Can you make this 2.8.12 compatible? This is what we have ob Ubuntu 14.04 LTS, which is available in the CI

**Comment by ksanderon** [comment]
@webmaster128 : it should not be big problem ;) I will observe #962. I suppose your changes are more important, so should be accepted before mine.

File: configure.py [hunk 2]
```diff
@@ -1420,6 +1424,106 @@ def bakefile_test_sources(file, sources):
...
+                for lib in using_mod.libs[options.os]:
+                    rules['libs'].add(lib)
+            if options.os in using_mod.frameworks:
+                for framework in using_mod.frameworks[options.os]:
+                    rules['frameworks'].add(framework)
+
+    def generate_sources_list(fd, name, rules):
+        fd.write('set(' + name + '\n')
+        for source in rules['sources']:
+            fd.write('    ${CMAKE_CURRENT_LIST_DIR}/' + source[2:] + '\n')
```
Comment (webmaster128) [inline configure.py]: Cutting substrings from paths looks dangerous to me. I think [os.path.normpath](https://docs.python.org/2/library/os.path.html#os.path.normpath) is your friend

```
$ python -c "import os; print(os.path.normpath('.'))"
.
$ python -c "import os; print(os.path.normpath('ab.cpp'))"
ab.cpp
$ python -c "import os; print(os.path.normpath('./sub/ab.cpp'))"
sub/ab.cpp

However, this will result in backward slashes on Windows and I don't know if this causes problems with cmake

File: configure.py [hunk 3]
```diff
@@ -1420,6 +1424,106 @@ def bakefile_test_sources(file, sources):
...
+
+    def generate_special_properties(fd, rules):
+        for source in rules['sources']:
+            joined = ' '.join(rules['sources'][source]['isa_flags'])
+            if joined:
+                fd.write('set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/' + source[2:] + ' PROPERTIES COMPILE_FLAGS '+ joined + ')\n')
+
+    lib_module = create_basic_rules(build_config.sources)
+    test_module = create_basic_rules(build_config.test_sources)
+    cli_module = create_basic_rules(build_config.cli_sources)
```
Comment (webmaster128) [inline configure.py]: You are mixing the terms "rules" and "module" here which makes it very hard to follow the code flow even for an experienced cmake/Python coder. Isn't the proper term you are searching for "target" (which is what cmake uses to describe a static/shared library or an executable)?

File: configure.py [hunk 4]
```diff
@@ -1420,6 +1424,106 @@ def bakefile_test_sources(file, sources):
...
+        fd.write('set(' + name + '\n')
+        for source in rules['sources']:
+            fd.write('    ${CMAKE_CURRENT_LIST_DIR}/' + source[2:] + '\n')
+        fd.write(')\n\n')
+
+    def generate_special_properties(fd, rules):
+        for source in rules['sources']:
+            joined = ' '.join(rules['sources'][source]['isa_flags'])
+            if joined:
+                fd.write('set_source_files_properties(${CMAKE_CURRENT_LIST_DIR}/' + source[2:] + ' PROPERTIES COMPILE_FLAGS '+ joined + ')\n')
```
Comment (webmaster128) [inline configure.py]: same here (normpath)

**Comment by webmaster128** [comment]
You might want to have a look at https://github.com/webmaster128/botan/commits/single-cmake (newest commit only), where I tried to do this thing in a static fashion some time ago. Please just copy/past of you find useful blocks in there. It's a dead branch an I am not planning to work on it on my own anymore.

File: configure.py (hunk 1 above)
Comment (ksanderon) [inline configure.py]: Yes, It should be not problem - I need only check if in 2.8.* are avaliable: 
* set_source_files_properties, 
* target_link_libraries,
* target_compile_options,
* target_include_directories,
* set_target_properties

As these are very useful to control target dependencies(by "smart" inheritance).

File: configure.py (hunk 3 above)
Comment (ksanderon) [inline configure.py]: Yes, I agree code is unclear here ;)

File: configure.py (hunk 4 above)
Comment (ksanderon) [inline configure.py]: Thank you for hint. I'll fix it.

**Comment by ksanderon** [comment]
merged and rebased

File: configure.py [hunk 5]
```diff
@@ -1429,6 +1433,120 @@ def bakefile_test_sources(file, sources):
...
+
+    def generate_target_sources_list(fd, targetName, target):
+        fd.write('set(' + targetName + '\n')
+        for source in target['sources']:
+            fd.write('    ${CMAKE_CURRENT_LIST_DIR}' + os.sep + os.path.normpath(source) + '\n')
+        fd.write(')\n\n')
+
+    def generate_target_source_files_isa_properties(fd, target):
+        for source in target['sources']:
+            joinedIsaFlags = ' '.join(target['sources'][source]['isa_flags'])
```
Comment (webmaster128) [inline configure.py]: We use snake_case [for variable names in Python](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables)

File: configure.py [hunk 6]
```diff
@@ -1429,6 +1433,120 @@ def bakefile_test_sources(file, sources):
...
+                    libs_or_frameworks_needed = True
+                    for isa in using_mod.need_isa:
+                        target['sources'][source_path]['isa_flags'].add(cc.isa_flags[isa])
+        if libs_or_frameworks_needed == True:
+            if options.os in using_mod.libs:
+                for lib in using_mod.libs[options.os]:
+                    target['libs'].add(lib)
+            if options.os in using_mod.frameworks:
+                for framework in using_mod.frameworks[options.os]:
+                    target['frameworks'].add('"-framework '+framework+'"')
```
Comment (webmaster128) [inline configure.py]: Using a string formatter makes it easier to understand what the final content will contain, especially in this case with all the quotes:

```python
target['frameworks'].add('"-framework %s"' % framework)
```

Comment (webmaster128) [inline configure.py]: And then I doubt that passing "-framework XYZ" including the quotes as a single command line argument is correct. Does that work?

File: configure.py [hunk 7]
```diff
@@ -1429,6 +1433,120 @@ def bakefile_test_sources(file, sources):
 
     f.close()
 
+def gen_cmake(build_config, using_mods, cc, options):
```
Comment (webmaster128) [inline configure.py]: Could you rename `build_config` to `build_paths` (for new code only)? I made the class more specific, such that it is easier to understand what is in there

File: configure.py [hunk 8]
```diff
@@ -1429,6 +1433,120 @@ def bakefile_test_sources(file, sources):
...
+                        target['sources'][source_path]['isa_flags'].add(cc.isa_flags[isa])
+        if libs_or_frameworks_needed == True:
+            if options.os in using_mod.libs:
+                for lib in using_mod.libs[options.os]:
+                    target['libs'].add(lib)
+            if options.os in using_mod.frameworks:
+                for framework in using_mod.frameworks[options.os]:
+                    target['frameworks'].add('"-framework '+framework+'"')
+
+    def generate_target_sources_list(fd, targetName, target):
```
Comment (webmaster128) [inline configure.py]: snake_case

Comment (ksanderon) [inline configure.py]: Power of habit :/ I'm afraid that, I will soon write mails with camelCase ;) Anyway, I will fix it.

File: configure.py (hunk 6 above)
Comment (ksanderon) [inline configure.py]: yes - this is somehow workaround for cmake. Without quotes it think that `-framework` `XYZ` are separate arguments ;) I'm not sure if this is cleanest solution but at least work correctly (generated  by cmake build commands are correct at least for make and ninja).

File: configure.py (hunk 7 above)
Comment (ksanderon) [inline configure.py]: done

File: configure.py (hunk 5 above)
Comment (ksanderon) [inline configure.py]: done

**Comment by randombit** [comment]
This looks ok to me, ready to merge?

**Comment by webmaster128** [comment]
👍 @randombit

File: configure.py [hunk 9]
```diff
@@ -1429,6 +1433,121 @@ def bakefile_test_sources(file, sources):
...
+                    libs_or_frameworks_needed = True
+                    for isa in using_mod.need_isa:
+                        target['sources'][source_path]['isa_flags'].add(cc.isa_flags[isa])
+        if libs_or_frameworks_needed == True:
+            if options.os in using_mod.libs:
+                for lib in using_mod.libs[options.os]:
+                    target['libs'].add(lib)
+            if options.os in using_mod.frameworks:
+                for framework in using_mod.frameworks[options.os]:
+                    target['frameworks'].add('"-framework %"' % framework)
```
Comment (webmaster128) [inline configure.py]: pylint just told me that there an "s" missing in this place: "-framework %" -> "-framework %s"

I am already fixing this and will submit erlier than expected because of this. But the version on master is not working
